### PR TITLE
Simplify extract_metadata.py by moving logic in tools/webassembly.py. NFC

### PR DESCRIPTION
**Pull request by sbc100**
This creates a caching layer and some extra helper functions on the
module object which avoid the need to track start such as the number
of imports elements of a given type.

### CODE REVIEW

File: tools/webassembly.py [hunk 1]
```diff
@@ -251,6 +263,8 @@ def parse_features_section(self):
     return features
 
   def parse_dylink_section(self):
+    if self._have_cached_section_info('dylink'):
+      return self._get_cached_section_info('dylink')
```
Comment (sbc100) [inline tools/webassembly.py:267]: I should probably turn this into a decorator... but maybe as a followup.

Comment (sbc100) [reply tools/webassembly.py:267]: Oh wait.. I can just use @functools memoization instead of all this.

File: tools/webassembly.py [hunk 2]
```diff
@@ -434,6 +443,37 @@ def get_tables(self):
   def has_name_section(self):
     return self.get_custom_section('name') is not None
 
+  _done_read_imports = False
```
Comment (dschuff) [inline tools/webassembly.py:446]: can this just be turned into `@cache` too?

Comment (sbc100) [reply tools/webassembly.py:446]: Its seems like a slightly different use case.. since it doesn't return anything.   But in theory yes I think so..

Comment (dschuff) [reply tools/webassembly.py]: Ah, right, it would be a bit weird. Well, either way is fine; I just noticed that it was asymmetric.

**Review by dschuff — APPROVED** [review]

**Comment by sbc100** [comment]
Seems that `once` is only in 3.9 .. fixed.. and created @once too.

File: tools/webassembly.py [hunk 3]
```diff
@@ -54,6 +54,28 @@ def read_sleb(iobuf):
   return leb128.i.decode_reader(iobuf)[0]
 
 
+# TODO(sbc): Use the builtin functools.cache once we update to python 3.9
```
Comment (dschuff) [inline tools/webassembly.py:57]: :disappointed:

**Comment by dschuff** [comment]
LGTM

**Review by kripken — APPROVED** [review]

**Pull request merged**